src/olympia/editors/management/commands/auto_approve.py

```diff
@@ -59,57 +66,77 @@ def handle(self, *args, **options):
-                log.info(
-                    'Version %s was skipped either because it had no '
-                    'file or because it had no validation attached.', version)
-                stats['error'] += 1
-            finally:
-                clear_reviewing_cache(version.addon.pk)
-
-        self.log_final_summary(stats, dry_run=dry_run)
-
-    def log_final_summary(self, stats, dry_run=False):
+            self.process(version)
+
+        self.log_final_summary(self.stats)
+
+    @transaction.atomic
+    def process(self, version):
+        """Process a single version, figuring out if it should be auto-approved
+        and calling the approval code if necessary."""
+
+        # Is the addon already locked by a reviewer ?
+        if get_reviewing_cache(version.addon.pk):
+            self.stats['locked'] += 1
+            return
+
+        # If admin review or more information was requested, skip this
+        # version, let a human handle it.
+        if version.addon.admin_review or version.has_info_request:
+            self.stats['flagged'] += 1
+            return
+
+        # Lock the addon for ourselves, no reviewer should touch it.
+        set_reviewing_cache(version.addon.pk, settings.TASK_USER_ID)
+
+        try:
+            log.info('Processing %s version %s...',
+                     unicode(version.addon.name), unicode(version.version))
+            summary, info = AutoApprovalSummary.create_summary_for_version(
+                version, max_average_daily_users=self.max_average_daily_users,
+                min_approved_updates=self.min_approved_updates,
+                dry_run=self.dry_run)
+            log.info('Auto Approval for %s version %s: %s',
+                     unicode(version.addon.name),
+                     unicode(version.version),
+                     summary.get_verdict_display())
+            self.stats.update({k: int(v) for k, v in info.items()})
+            if summary.verdict == self.successful_verdict:
+                self.stats['auto_approved'] += 1
+                if summary.verdict == amo.AUTO_APPROVED:
+                    self.approve(version)
+
+        except (AutoApprovalNotEnoughFilesError,
+                AutoApprovalNoValidationResultError):
+            log.info(
+                'Version %s was skipped either because it had no '
+                'file or because it had no validation attached.', version)
+            self.stats['error'] += 1
+        finally:
+            clear_reviewing_cache(version.addon.pk)
+
+    def approve(self, version):
+        """Do the approval itself, caling ReviewHelper to change the status,
+        sign the files, send the e-mail, etc."""
+        # Note: this should automatically use the TASK_USER_ID user.
+        helper = ReviewHelper(addon=version.addon, version=version)
+        helper.handler.data = {
+            'comments': _(u'This version has been approved for the public.'
+                          u'\r\n\r\nThank you!')
```

Comment:
I stole this from one of the `CannedResponse`s, and it had `\r\n`, so I kept them. (I couldn't easily fetch and use the `CannedResponse` programmatically as there is no way to identify it)